modules/terraform/validate_struct.go

```diff
@@ -0,0 +1,110 @@
+	// RootDir is /home/project and you want to include everything EXCEPT /home/project/modules, add
+	// /home/project/modules to this slice. Note that ExcludeDirs is only considered when IncludeDirs is not passed
+	// Note that while the struct requires full paths, you can pass relative paths to the NewValidationOptions function
+	// which will build the full paths based on the supplied RootDir
+	ExcludeDirs []string
+}
+
+// NewValidationOptions returns a ValidationOptions struct, with override-able sane defaults. Note that the
+// ValidationOptions's fields IncludeDirs and ExcludeDirs must be full paths, but this method will accept relative paths
+// and build the full paths when instantiating the ValidationOptions struct,  making it the preferred means of configuring
+// ValidationOptions.
+//
+// For example, if your RootDir is /home/project/ and you want to exclude "modules" and "test" you need
+// only pass the relative paths in your excludeDirs slice like so:
+// opts, err := NewValidationOptions("/home/project", []string{}, []string{"modules", "test"})
+func NewValidationOptions(rootDir string, includeDirs, excludeDirs []string) (ValidationOptions, error) {
+	vo := ValidationOptions{
+		RootDir:     "",
+		IncludeDirs: []string{},
+		ExcludeDirs: []string{},
+	}
+
+	if rootDir == "" {
+		return vo, ValidationUndefinedRootDirErr{}
+	}
+
+	vo.RootDir = rootDir
+
+	if len(includeDirs) > 0 {
+		vo.IncludeDirs = buildFullPathsFromRelative(vo.RootDir, includeDirs)
+	}
+
+	if len(excludeDirs) > 0 {
+		vo.ExcludeDirs = buildFullPathsFromRelative(vo.RootDir, excludeDirs)
+	}
+
+	return vo, nil
+}
+
+func buildFullPathsFromRelative(rootDir string, relativePaths []string) []string {
+	var fullPaths []string
+	for _, relativePath := range relativePaths {
+		fullPaths = append(fullPaths, filepath.Join(rootDir, relativePath))
+	}
+	return fullPaths
+}
+
+// FindTerraformModulePathsInRootE returns a slice strings representing the filepaths for all valid Terraform modules
+// in both the "modules" directory and "examples" directories in the project root, if they exist.
+func FindTerraformModulePathsInRootE(opts ValidationOptions) ([]string, error) {
+	// Find all Terraform files from the configured RootDir
+	pattern := fmt.Sprintf("%s/**/*.tf", opts.RootDir)
+	matches, err := zglob.Glob(pattern)
+	if err != nil {
+		return matches, err
+	}
+	// Keep a unique set of the base dirs that contain Terraform files
+	terraformDirSet := make(map[string]bool)
+	for _, match := range matches {
+		// The glob match returns all full paths to every .tf file, whereas we're only interested in their root
+		// directories for the purposes of running Terraform validate
+		rootDir := path.Dir(match)
+		terraformDirSet[rootDir] = true
+	}
+
+	// Return the unique slice of Terraform directories found starting at opts.RootDir
+	terraformDirs := make([]string, 0, len(terraformDirSet))
+	for dir := range terraformDirSet {
+		terraformDirs = append(terraformDirs, dir)
+	}
```

Comment:
Is this not the same as:

```go
terraformDirs := keys(terraformDirSet)
```

Reply:
I was looking for this `keys` method before I settled on the above implementation, but my current understanding is Go doesn't support this. What am I missing?

Reply:
Oh, sorry, brain fart. I thought Go had a native `keys` method.

`go-commons` does have this: https://github.com/gruntwork-io/go-commons/blob/master/collections/maps.go#L19. Not sure we use `go-commons` in Terratest yet?

Reply:
Seems we do make liberal use of it - so cut over to it in 70e6c2d. Required slight non-material modification of the map type to satisfy the method signature of `Keys`.